ui/src/dashboards/components/GraphOptionsCustomizableField.tsx

```diff
@@ -1,70 +1,186 @@
 import React, {PureComponent} from 'react'
+import {findDOMNode} from 'react-dom'
+import {
+  DragSourceSpec,
+  DropTargetConnector,
+  DragSourceMonitor,
+  DragSource,
+  DropTarget,
+  DragSourceConnector,
+} from 'react-dnd'
 
-import InputClickToEdit from 'src/shared/components/InputClickToEdit'
+const FieldType = 'field'
 
 interface Field {
   internalName: string
   displayName: string
   visible: boolean
+  order?: number
+  index?: number
 }
 
-interface Props {
+interface FieldSourceProps {
   internalName: string
   displayName: string
   visible: boolean
-  onFieldUpdate: (field: Field) => void
+  index: number
+  id: string
+  key: string
+  isDragging?: boolean
+  onFieldUpdate?: (field: Field) => void
+  connectDragSource?: any
+  connectDropTarget?: any
+  connectDragPreview?: any
+  moveField: (dragIndex: number, hoverIndex: number) => void
 }
 
-class GraphOptionsCustomizableField extends PureComponent<Props, {}> {
+const fieldSource: DragSourceSpec<FieldSourceProps> = {
+  beginDrag(props) {
+    return {
+      id: props.id,
+      index: props.index,
+    }
+  },
+}
+
+const fieldTarget = {
+  hover(props, monitor, component) {
+    const dragIndex = monitor.getItem().index
+    const hoverIndex = props.index
+
+    // Don't replace items with themselves
+    if (dragIndex === hoverIndex) {
+      return
+    }
+
+    // Determine rectangle on screen
+    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect()
+
+    // Get vertical middle
+    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2
+
+    // Determine mouse position
+    const clientOffset = monitor.getClientOffset()
+
+    // Get pixels to the top
+    const hoverClientY = clientOffset.y - hoverBoundingRect.top
+
+    // Only perform the move when the mouse has crossed half of the items height
+    // When dragging downwards, only move when the cursor is below 50%
+    // When dragging upwards, only move when the cursor is above 50%
+
+    // Dragging downwards
+    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+      return
```

Comment:
instead of checking when not to do the move and returning, you could instead just check when to move and perform the move there

Reply:
I feel like this works just as well.